### PR TITLE
Handle default ports when the host header has no port specified

### PR DESCRIPTION
**Pull request by wkirschbaum**
Cowboy will automatically default to port 80 when no port has been given, but Bandit does not. Bandit will always default to the underlying port, which makes it hard to know if there was a port specified or if a port was assumed.

This change ignores cowboy or bandit's assumption around port number and parses it from the header directly.

This solves: https://github.com/ueberauth/ueberauth/issues/180

### CODE REVIEW

**Comment by wkirschbaum** [comment]
I would still like to write a test for this.

File: lib/ueberauth/strategies/helpers.ex [hunk 1]
```diff
@@ -241,14 +242,18 @@ defmodule Ueberauth.Strategy.Helpers do
...
+        |> List.first()
+      [host | _] ->
+        host
+    end
   end
 
-  defp normalize_port("https", 80), do: 443
-  defp normalize_port(_, port) when is_integer(port), do: port
+  defp normalize_port("https", nil), do: 443
+  defp normalize_port(_, nil), do: 80
```
Comment (mtrudel) [inline lib/ueberauth/strategies/helpers.ex:256]: `URI` provides a function for exactly this: https://hexdocs.pm/elixir/1.12/URI.html#default_port/1

Comment (wkirschbaum) [reply lib/ueberauth/strategies/helpers.ex:256]: correct, i saw it just an hour ago and wanted to update the code 😆. Will add a test or two and update to use the URI function

**Comment by yordis** [comment]
Hey, do you mind following https://github.com/ueberauth/ueberauth/blob/master/CONTRIBUTING.md#pull-requests-welcome and updating the changelog and whatnot?

**Comment by wkirschbaum** [comment]
@yordis i assume 6 and 7? happy to.

**Review by yordis — APPROVED** [review]

**Comment by wkirschbaum** [comment]
I pushed a mix format change

**Comment by jamesvl** [comment]
So it looks like this change has the side effect of breaking "typical" Phoenix setups which might run on port 4000 (http) or 4001 (https).

By removing `conn.port` (which I understand was part of the goal) from `normalize_port(scheme, port || conn.port)` a local developer instance will break unless `callback_port: 4001` is explicitly added in the `dev.exs` config for providers that use callback. (the `Identity` provider will not have issues)

**Comment by jamesvl** [comment]
We should probably remove this line from the docs now:

> By default, Ueberauth uses your [Plug.Conn](https://hexdocs.pm/plug/1.14.0/Plug.Conn.html) port.

**Comment by mtrudel** [comment]
It's a bit of a tricky thing. See my recent upstream-ish comment at https://github.com/mtrudel/bandit/issues/106#issuecomment-1458847902 for my framing of the root problem here. 

tl;dr: there's no clear consensus what `Plug.Conn.port` refers to, whether it's 'what the user thinks they're connected to' or 'what is the actual port the connection is coming in on'. This becomes salient when e.g. the server is running behind a load balancer.

**Comment by wkirschbaum** [comment]
@jamesvl I just tested the scenarios I believe you describe both locally and on a staging deploy behind a load balancer and see not issue.   We are still falling back to the Plug.Conn port https://github.com/ueberauth/ueberauth/blob/master/lib/ueberauth/strategies/helpers.ex#L216 and only use a default port if there is none: https://github.com/ueberauth/ueberauth/blob/master/lib/ueberauth/strategies/helpers.ex#L256. 

So really understanding the issue.

**Comment by jamesvl** [comment]
The line you highlighted is getting the `conn.host`, not the `conn.port`.

The scenario I'm referring to is for local development, when there is no load balancer or reverse proxy and the browser goes directly to port 4000 or 4001 with Phoenix + Cowboy. (that's the "default" I was trying to refer to above)

**Comment by jamesvl** [comment]
I agree with the confusion in the Bandit issue - `conn.port` refers only to what Phoenix is listening on and may or may not have anything to do with the actual URL of the app. (That's what `Phoenix.Endpoint` config for `url` is supposed to specify, but Ueberauth has no clean way to pull the Endpoint config, and I don't think it should.)

I don't know if I have a clean way to resolve this ambiguity. Locally, I've adapted logic to the following, which seems to work:
* if there is a `x-forwarded-host` or a `host` header present, use that for `host`; otherwise pull from `conn.host`
* if either of the above `host` headers is present, check to see if there is a `port` value present and use it
* otherwise, `port` comes from
  * `conn.port` when *no* host header is present
  * the Ueberauth explicit config (if present) and lastly from `URI.default_port/1`


That's a bit convoluted I know, but it would pass all the following scenarios:
* new Phoenix + Cowboy install, with minimal or no port customization anywhere
* the `curl` commands that set the `host` header (e.g. what Bandit is doing?)
* getting the port from the Ueberauth config whenever it is explicitly given
* using the port from a reverse proxy or load balancer that sets `x-forwarded-host`

**Comment by jamesvl** [comment]
Here's a branch with what I've done locally, if it help clarify the above: https://github.com/ueberauth/ueberauth/compare/master...Semsee:ueberauth:feature/respect-host-header

**Comment by wkirschbaum** [comment]
The line below splits out the port. 

> The line you highlighted is getting the `conn.host`, not the `conn.port`.
> 
> The scenario I'm referring to is for local development, when there is no load balancer or reverse proxy and the browser goes directly to port 4000 or 4001 with Phoenix + Cowboy. (that's the "default" I was trying to refer to above).
> 
> Cowboy is not sending `host` header (I assume Bandit is?) so I think we're seeing two different things locally.

My mistake, we default to the port for conn.host, not to conn.port.

Are you sure it does not work locally? I just tested on two projects for both bandit and cowboy and they both work as expected.  Is there perhaps a unit test to illustrate the issue?

As soon as we use conn.port we get an inconsistent map some port based on the adapter.

**Comment by jamesvl** [comment]
I could write a unit test for it, but I would need to bake into it some assumption about the meaning of `conn.port` when no `host` header is present... and that's the heart of the issue.

For your two test projects, are you running Cowboy on port 4000 and 4001? If so, how is Ueberauth picking up that the redirect URL should be to that port? Maybe it's something simple I'm missing in my config.

**Comment by wkirschbaum** [comment]
Its running on port 4000, stock standard install. The host header will have the port and we pull it from the host header, same as what cowboy and bandit does, except the logic is different for what to do with the port. I am not sure what you mean by port 4000 and 4001, if i change the endpoint port to 4001 it works as well ( given the idp accepts it ).

Your proposed breaks 6 unit tests.

I am missing something here, sorry.

**Comment by jamesvl** [comment]
Yep, I'm checking why the `host` header isn't getting picked up for me locally - maybe this is something my Plug pipeline is dropping? If it is, then this change should be fine...

(Sorry - I know the unit tests aren't passing on my branch - I was putting up code only to illustrate what English was doing a poor job of describing.)

**Comment by wkirschbaum** [comment]
Phoenix 1.6 and 1.7 with their latest respected cowboy versions. I doubt phoenix has anything to do with it though. cowboy and bandit builds the conn and ueberauth uses it directly.

**Comment by jamesvl** [comment]
Yep, I have *no* idea why `host` is getting dropped from my `conn.req_headers` (it's a required header for http!)... I see nothing in my code base that would drop, merge, or overwrite r`eq_headers`...

I guess it's fine, and something weird about my setup.

**Comment by jamesvl** [comment]
I think I've narrowed it down, even using a brand new Phoenix 1.7 project and either Cowboy or Bandit.

Run `mix phx.gen.cert` then add the following to the `Endpoint` config in `config/dev.exs` for Bandit:

```elixir
  https: [
    port: 4040,
    cipher_suite: :strong,
    transport_options: [
      certfile: "priv/cert/selfsigned.pem",
      keyfile: "priv/cert/selfsigned_key.pem"
    ]
  ]
```

(You may use the same config for Cowboy except `certfile` and `keyfile` are not nested under `transport_options`.)

Restart the app; when I visit port 4000 I *do* see `{"host", "localhost:4000"}` in the `conn.req_headers`. When I visit port 4040 it is missing. @wkirschbaum does this replicate for you?

**Comment by achedeuzot** [comment]
Hi @jamesvl and @wkirschbaum 
I've stumbled on the same issue as @jamesvl is describing in local development mode (phoenix running on my machine with no reverse proxy, port 4000, SSL enabled) after upgrading `ueberauth` to the latest version (from 0.10.3 to 0.10.5). When doing the initial oauth2 call, the port is now missing in the `redirect_uri` where it previously wasn't missing. The host I have is `https://127.0.0.1:4000/...` and since the change, it's using `https://127.0.0.1/` (notice the port has been dropped; I'm also using httpS). So the development workflow auth is broken with this change because the callback URL is wrong.
Have you solved the issue since the last exchanges on this ticket ? :)

**Comment by wkirschbaum** [comment]
@achedeuzot see: https://github.com/ueberauth/ueberauth/pull/187
